File: Vigenère-and-Gronsfeld-Cipher.java
import java.util.Scanner;

public class MyProgram{
	
	//declare the tableau 2D char-array and string keyword variable
	public static char[][] tableau = new char[26][26];
	public static String keyword = new String("");
	
	static void makeTableau() {
		//creating the tableau
		for(int i=0;i<26;i++) {
			for(int j=i;j<i+26;j++) {
				int a = j+65;
				if (a>90) {
					tableau[i][j-i] = (char)(a%90+64);
				}
				else {
					tableau[i][j-i] = (char)a;
				}
			}
		}
	}
	
	static void makeKeyword(int n) {
		//creating the keyword
		char key[] = {'F','O','R','T','I','F','I','C','A','T','I','O','N'};
		for(int i=0;i<n;i++) {
			keyword = keyword + Character.toString(key[i%13]);
		}
	}
	
	//function for encryption
	static String encrypt(String pt) {
		char[] encpt1 = pt.toCharArray();
		char[] charArrayKeyword = keyword.toCharArray();
		String encpt2 = new String("");
		for(int i=0;i<pt.length();i++) {
			int j=charArrayKeyword[i];
			int a = j-65;
			int b = encpt1[i];
			b = b-65;
			encpt2 = encpt2 + Character.toString(tableau[a][b]);
			}
		return encpt2;
	}
	
	//function for decryption
	static String decrypt(String encpt) {
		char[] decpt1 = keyword.toCharArray();
		char[] decpt2 = encpt.toCharArray();
		String decpt3 = new String("");
		for(int k=0; k<keyword.length(); k++) {
			int i=decpt1[k],j,b=0;
			i = i-65;
			for(j=0;j<26;j++) {
				if (tableau[i][j]==decpt2[k]) {
					b=j;
					break;
				}
			}
			decpt3 = decpt3 + Character.toString((char)(b+65));
		}
		return decpt3;
	}
	
	//main method
	public static void main(String[] args) {
		
		makeTableau();
		System.out.print("Enter plaintext: ");
		Scanner scan = new Scanner(System.in);
		String pt = scan.nextLine();
		makeKeyword(pt.length());
		String encpt = encrypt(pt);
		System.out.println("Encrypted text: "+encpt);
		String decpt = decrypt(encpt);
		System.out.println("Decrypted text: "+decpt);
		
	}
}

File: Vigenère-and-Gronsfeld-Cipher.cpp
#include<iostream>
#include<string.h>
#include<string>
#include <bits/stdc++.h>

char tableau[26][26];
char key[] = {'F','O','R','T','I','F','I','C','A','T','I','O','N'};
std::string keyword("");

void makeTableau()
{
	int i,j;
	for(i=0;i<26;i++)
		for(j=i;j<i+26;j++)
		{
			
			int a = j+65;
			if(a>90)
			{
				char c = a%91+65;
				tableau[i][j-i] = c;
			}
			else
			{
				char c = a;
				tableau[i][j-i] = c;
			}
		}
}

void makeKeyword(int n)
{
	int i;
	for(i=0;i<n;i++)
	{
		std::string s(1,key[i%n]);
		keyword = keyword + s;
	}
}

std::string encrypt(std::string pt)
{
	std::string encpt = "";
	char pt_char[pt.length()+1];
	strcpy(pt_char,pt.c_str());
	char kw_char[keyword.length()+1];
	strcpy(kw_char,keyword.c_str());
	int i,j;
	for(int a=0;a<pt.length();a++)
	{
		i = kw_char[a];
		i = i-65;
		j = pt_char[a];
		j = j-65;
		std::string s(1,tableau[i][j]);
		encpt = encpt + s;	
	}
	return encpt;
}

std::string decrypt(std::string encpt)
{
	std::string decpt = "";
	char encpt_char[encpt.length()+1];
	strcpy(encpt_char,encpt.c_str());
	char kw_char[keyword.length()+1];
	strcpy(kw_char,keyword.c_str());
	for(int a=0;a<encpt.length();a++)
	{
		int i = kw_char[a];
		i-=65;
		for(int b=0;b<26;b++)
		{
			if(tableau[i][b]==encpt_char[a])
			{
				char c = b+65;
				std::string s(1,c);
				decpt = decpt + s;
				break;
			}
		}
	}
	return decpt;
}

int main()
{
	std::string pt;
	std::cout<<"Enter plaintext: ";
	std::cin>>pt;
	makeTableau();
	makeKeyword(pt.length());
	std::string encpt = encrypt(pt);
	std::cout<<"Encrypted text: "<<encpt<<"\n";
	std::string decpt = decrypt(encpt);
	std::cout<<"Decrypted text: "<<decpt<<"\n";
	return 0;
}

